Content.Server/Traits/Assorted/HeirloomSystem.cs

```diff
@@ -0,0 +1,39 @@
+using System.Linq;
+using Content.Shared.Mood;
+using Content.Shared.Traits.Assorted.Components;
+
+namespace Content.Server.Traits.Assorted;
+
+public sealed class HeirloomSystem : EntitySystem
+{
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        //TODO: This could probably use some optimization
+        var query = EntityManager.EntityQueryEnumerator<HeirloomHaverComponent>();
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            var children = RecursiveGetAllChildren(uid);
+            if (!children.Any(c => c == comp.Heirloom))
+                continue;
+            var ev = new MoodEffectEvent(comp.Moodlet);
+            RaiseLocalEvent(uid, ev);
+        }
+
+        query.Dispose();
+    }
+
+    /// A reasonable assumption
+    private IEnumerable<EntityUid> RecursiveGetAllChildren(EntityUid uid)
+    {
+        var xform = Transform(uid);
+        var children = xform.ChildEnumerator;
+        while (children.MoveNext(out var child))
+        {
+            yield return child;
```

Comment:
what is a yield return!?!!?

Reply:
it stops the execution of the program, after returning the result, and whenever you run the function again, it will resume on the control flow where it left off https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/yield

used very extensively in sharedbodysystem if you care